codecs/h264_packet.go

```diff
@@ -127,6 +127,144 @@ func emitNalus(nals []byte, emit func([]byte)) {
 	}
 }
 
+func (p *H264Payloader) PayloadAndDetectData(mtu uint16, payload []byte) ([][]byte, int) {
+	var payloads [][]byte
+	dataPayloadIdx := -1
+	dataPayloadFound := false
+	payloadIdx := 0
+
+	if len(payload) == 0 {
+		return payloads, dataPayloadIdx
+	}
+
+	emitNalus(payload, func(nalu []byte) {
+		if len(nalu) == 0 {
+			return
+		}
+
+		naluType := nalu[0] & naluTypeBitmask
+		naluRefIdc := nalu[0] & naluRefIdcBitmask
+
+		switch {
+		case naluType == audNALUType || naluType == fillerNALUType:
+			return
+		case naluType == spsNALUType:
+			p.spsNalu = nalu
+			return
+		case naluType == ppsNALUType:
+			p.ppsNalu = nalu
+			return
+		case p.spsNalu != nil && p.ppsNalu != nil:
+			// Pack current NALU with SPS and PPS as STAP-A
+			spsLen := make([]byte, 2)
+			binary.BigEndian.PutUint16(spsLen, uint16(len(p.spsNalu)))
+
+			ppsLen := make([]byte, 2)
+			binary.BigEndian.PutUint16(ppsLen, uint16(len(p.ppsNalu)))
+
+			stapANalu := []byte{outputStapAHeader}
+			stapANalu = append(stapANalu, spsLen...)
+			stapANalu = append(stapANalu, p.spsNalu...)
+			stapANalu = append(stapANalu, ppsLen...)
+			stapANalu = append(stapANalu, p.ppsNalu...)
+			if len(stapANalu) <= int(mtu) {
+				out := make([]byte, len(stapANalu))
+				copy(out, stapANalu)
+				payloads = append(payloads, out)
+				payloadIdx++
+			}
+
+			p.spsNalu = nil
+			p.ppsNalu = nil
+		}
+
+		// Single NALU
+		if len(nalu) <= int(mtu) {
+			out := make([]byte, len(nalu))
+			copy(out, nalu)
+			payloads = append(payloads, out)
+			// still looking for first packet of data
+			if !dataPayloadFound {
+				if naluType == 1 || naluType == 2 || naluType == 5 { // this is a data nalu
+					dataPayloadFound = true
+					dataPayloadIdx = payloadIdx
```

Comment:
You don't really need to track an index counter. the data payload was just appended, so
dataPayloadIdx = len(payloads)-1